File: backend/app/schemas/interview.py
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.audio import TranscribeData
from app.schemas.evaluation import EvaluateData, ReportData
from app.schemas.resume import CandidateProfile


class StartInterviewRequest(BaseModel):
    resume_id: Optional[UUID] = None
    interview_type: Literal[
        "sde",
        "sdet",
        "backend",
        "behavioral",
        "system_design",
        "resume_based",
        "jd_based",
    ]
    difficulty: Literal["easy", "medium", "hard"] = "medium"
    job_description: Optional[str] = None
    target_company: Optional[str] = None
    target_role: Optional[str] = None
    question_count: int = Field(default=5, ge=1, le=5)
    voice_enabled: bool = True


class SessionLimits(BaseModel):
    max_questions: int
    max_answer_duration_seconds: int


class SessionNextAction(BaseModel):
    type: str
    endpoint: str


class SessionData(BaseModel):
    session_id: UUID
    resume_id: Optional[UUID] = None
    interview_type: str
    difficulty: str
    target_role: Optional[str] = None
    target_company: Optional[str] = None
    voice_enabled: bool
    question_count: int
    status: str
    started_at: datetime
    limits: SessionLimits
    next_action: SessionNextAction
    expires_at: Optional[datetime] = None


class GenerateQuestionRequest(BaseModel):
    session_id: UUID
    mode: Literal["first", "next", "follow_up"] = "first"
    previous_answer_id: Optional[UUID] = None
    include_voice: bool = True
    question_count: int = Field(default=5, ge=1, le=5)


class QuestionData(BaseModel):
    question_id: UUID
    session_id: UUID
    sequence: int
    question_text: str
    question_type: str
    audio_url: Optional[str] = None
    voice_enabled: bool
    latency_state: Optional[dict] = None


class AudioData(BaseModel):
    enabled: bool
    provider: Literal["elevenlabs", "browser"] | None = None
    audio_url: Optional[str] = None
    duration_seconds: Optional[float] = None
    cached: bool = False
    label: str = ""
    upgrade_required: bool = False
    browser_speech_text: Optional[str] = None


class QuestionDetail(BaseModel):
    question_id: UUID
    sequence: int
    type: str
    difficulty: str
    question_text: str
    expected_focus_areas: list[str] = []
    time_limit_seconds: int = 60
    audio: AudioData


class LatencyState(BaseModel):
    current: str
    completed_steps: list[str] = []


class QuestionResponseData(BaseModel):
    session_id: UUID
    question: QuestionDetail
    latency_state: LatencyState


class HistoryItem(BaseModel):
    session_id: UUID
    interview_type: str
    target_role: Optional[str] = None
    target_company: Optional[str] = None
    status: str
    question_count: int
    overall_score: Optional[float] = None
    started_at: datetime
    ended_at: Optional[datetime] = None


class Pagination(BaseModel):
    page: int
    limit: int
    total_items: int
    total_pages: int


class HistoryData(BaseModel):
    items: list[HistoryItem]
    pagination: Pagination


class CompletedAnswerData(BaseModel):
    questionNumber: int
    questionId: str
    questionText: str
    questionType: str
    answerId: str
    transcript: str
    wordCount: int
    scores: dict
    overallScore: float | None = None
    feedbackSummary: str
    strengths: list[str] = []
    improvements: list[str] = []


class SessionDetailData(BaseModel):
    session_id: UUID
    resume_id: Optional[UUID] = None
    interview_type: str
    difficulty: str
    target_role: Optional[str] = None
    target_company: Optional[str] = None
    job_description: Optional[str] = None
    question_count: int
    voice_enabled: bool
    status: str
    started_at: datetime
    expires_at: Optional[datetime] = None
    questions_answered: int
    current_sequence: int
    last_activity_at: datetime
    resume_profile: Optional[CandidateProfile] = None
    current_question: Optional[QuestionDetail] = None
    current_transcript: Optional[TranscribeData] = None
    current_evaluation: Optional[EvaluateData] = None
    completed_answers: list[CompletedAnswerData] = []
    final_report: Optional[ReportData] = None

File: frontend/types/interview.ts
export type InterviewMode = "first" | "next" | "follow_up";
export type InterviewType =
  | "sde"
  | "sdet"
  | "backend"
  | "behavioral"
  | "system_design"
  | "resume_based"
  | "jd_based";
export type InterviewDifficulty = "easy" | "medium" | "hard";

export type LatencyStateType =
  | "idle"
  | "question_generating"
  | "voice_generating"
  | "ready_for_answer"
  | "recording_answer"
  | "uploading_audio"
  | "transcribing_answer"
  | "transcript_ready"
  | "evaluating_answer"
  | "preparing_follow_up"
  | "evaluation_ready"
  | "next_question_loading"
  | "interview_complete"
  | "final_report_generating"
  | "scorecard_ready"
  | "error";

export interface AudioData {
  enabled: boolean;
  provider: "elevenlabs" | "browser" | null;
  audio_url: string | null;
  duration_seconds: number | null;
  cached: boolean;
  label: string;
  upgrade_required: boolean;
  browser_speech_text: string | null;
}

export interface QuestionDetail {
  question_id: string;
  sequence: number;
  type: string;
  difficulty: string;
  question_text: string;
  expected_focus_areas: string[];
  time_limit_seconds: number;
  audio: AudioData;
}

export interface LatencyState {
  current: string;
  completed_steps: string[];
}

export interface QuestionResponseData {
  session_id: string;
  question: QuestionDetail;
  latency_state: LatencyState;
}

export interface ApiSuccessResponse<T> {
  success: true;
  data: T;
  message: string;
}

export interface ApiErrorResponse {
  success: false;
  error: {
    code: string;
    message: string;
    details?: Record<string, unknown>;
  };
}

export type ApiResponse<T> = ApiSuccessResponse<T> | ApiErrorResponse;

export interface GenerateQuestionPayload {
  session_id: string;
  mode: InterviewMode;
  include_voice: boolean;
  previous_answer_id?: string;
  question_count?: number;
}

export interface StartInterviewPayload {
  resume_id?: string | null;
  interview_type: InterviewType;
  difficulty: InterviewDifficulty;
  job_description?: string | null;
  target_company?: string | null;
  target_role?: string | null;
  question_count: number;
  voice_enabled: boolean;
}

export interface SessionLimits {
  max_questions: number;
  max_answer_duration_seconds: number;
}

export interface SessionNextAction {
  type: string;
  endpoint: string;
}

export interface StartInterviewResult {
  session_id: string;
  resume_id: string | null;
  interview_type: InterviewType;
  difficulty: InterviewDifficulty;
  target_role: string | null;
  target_company: string | null;
  voice_enabled: boolean;
  question_count: number;
  status: string;
  started_at: string;
  limits: SessionLimits;
  next_action: SessionNextAction;
  expires_at: string | null;
}

export interface FillerWords {
  count: number;
  examples: string[];
}

export interface TranscribeLatency {
  transcription_ms: number;
}

export interface TranscribeResult {
  answer_id: string;
  session_id: string;
  question_id: string;
  transcript: string;
  language: string;
  duration_seconds: number;
  word_count: number;
  filler_words: FillerWords;
  raw_audio_deleted: boolean;
  submitted_at: string;
  latency: TranscribeLatency;
}

export interface TranscribePayload {
  session_id: string;
  question_id: string;
  duration_seconds: number;
  language?: string;
  audio_file: Blob;
}

export interface EvaluationScores {
  technical_correctness: number | null;
  clarity: number | null;
  depth: number | null;
  confidence: number | null;
  relevance: number | null;
  structure: number | null;
  communication: number | null;
  conciseness: number | null;
  example_quality: number | null;
  overall: number | null;
}

export interface EvaluationFeedback {
  summary: string;
  strengths: string[];
  improvements: string[];
  ideal_answer_points: string[];
}

export interface FollowUp {
  recommended: boolean;
  question_text: string | null;
}

export interface EvaluationLatency {
  evaluation_ms: number;
}

export interface EvaluateResult {
  score_id: string;
  session_id: string;
  question_id: string;
  answer_id: string;
  scores: EvaluationScores;
  feedback: EvaluationFeedback;
  follow_up: FollowUp;
  latency: EvaluationLatency;
}

export interface EvaluatePayload {
  session_id: string;
  question_id: string;
  answer_id: string;
  generate_follow_up: boolean;
}

export interface CompletedAnswer {
  questionNumber: number;
  questionId: string;
  questionText: string;
  questionType: string;
  answerId: string;
  transcript: string;
  wordCount: number;
  scores: EvaluationScores;
  overallScore: number | null;
  feedbackSummary: string;
  strengths: string[];
  improvements: string[];
}

export interface InterviewSession {
  sessionId: string;
  maxQuestions: number;
  completedAnswers: CompletedAnswer[];
  isComplete: boolean;
}

export interface FinalScoreBreakdown {
  technical: number;
  communication: number;
  confidence: number;
  problem_solving: number;
  role_fit: number;
}

export interface FinalQuestionReview {
  question_id: string;
  sequence: number;
  question_text: string;
  answer_id: string;
  overall_score: number;
  feedback_summary: string;
}

export interface FinalTranscriptItem {
  question: string;
  answer: string;
}

export interface FinalReport {
  report_id: string;
  session_id: string;
  status: string;
  overall_score: number;
  score_breakdown: FinalScoreBreakdown;
  summary: string;
  strengths: string[];
  weaknesses: string[];
  recommended_topics: string[];
  question_reviews: FinalQuestionReview[];
  transcript: FinalTranscriptItem[] | null;
  created_at: string;
}

export interface FinalReportPayload {
  session_id: string;
  include_transcript: boolean;
  include_recommendations: boolean;
}

export interface SkillSet {
  languages: string[];
  frameworks: string[];
  databases: string[];
  cloud_devops: string[];
  testing_tools: string[];
  other: string[];
}

export interface ExperienceItem {
  company: string | null;
  role: string | null;
  start_date: string | null;
  end_date: string | null;
  responsibilities: string[];
  achievements: string[];
  technologies: string[];
}

export interface ProjectItem {
  name: string | null;
  description: string | null;
  technologies: string[];
  interview_focus: string[];
}

export interface ResumeProfile {
  candidate_name: string | null;
  email: string | null;
  current_or_latest_role: string | null;
  total_experience_years: number | null;
  summary: string | null;
  skills: SkillSet;
  experience: ExperienceItem[];
  projects: ProjectItem[];
  strength_areas: string[];
  recommended_interview_topics: string[];
}

export interface ResumeUploadResult {
  resume_id: string;
  file_name: string;
  parsed: boolean;
  profile: ResumeProfile | null;
  created_at: string;
}

export interface HistoryItem {
  session_id: string;
  interview_type: string;
  target_role: string | null;
  target_company: string | null;
  status: string;
  question_count: number;
  overall_score: number | null;
  started_at: string;
  ended_at: string | null;
}

export interface Pagination {
  page: number;
  limit: number;
  total_items: number;
  total_pages: number;
}

export interface HistoryData {
  items: HistoryItem[];
  pagination: Pagination;
}

export interface SessionDetailResult {
  session_id: string;
  resume_id: string | null;
  interview_type: InterviewType;
  difficulty: InterviewDifficulty;
  target_role: string | null;
  target_company: string | null;
  job_description: string | null;
  question_count: number;
  voice_enabled: boolean;
  status: string;
  started_at: string;
  expires_at: string | null;
  questions_answered: number;
  current_sequence: number;
  last_activity_at: string;
  resume_profile: ResumeProfile | null;
  current_question: QuestionDetail | null;
  current_transcript: TranscribeResult | null;
  current_evaluation: EvaluateResult | null;
  completed_answers: CompletedAnswer[];
  final_report: FinalReport | null;
}
